File: src-tauri/src/credentials.rs
use crate::error::AppError;
use keyring::Entry;
use serde::{Deserialize, Serialize};

const SERVICE_NAME: &str = "dev.xikxp1.claude-monitor";
const CREDENTIALS_KEY: &str = "credentials";
const OLLAMA_CREDENTIALS_KEY: &str = "ollama_credentials";

#[derive(Serialize, Deserialize)]
struct StoredCredentials {
    organization_id: String,
    session_token: String,
}

/// Load credentials from OS keychain.
/// Returns None if credentials don't exist or on any error.
pub fn load_credentials() -> Option<(String, String)> {
    let entry = Entry::new(SERVICE_NAME, CREDENTIALS_KEY).ok()?;
    let json = entry.get_password().ok()?;
    let creds: StoredCredentials = serde_json::from_str(&json).ok()?;
    Some((creds.organization_id, creds.session_token))
}

/// Save credentials to OS keychain.
pub fn save_credentials(org_id: &str, session_token: &str) -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, CREDENTIALS_KEY)
        .map_err(|e| AppError::Storage(format!("Failed to create keyring entry: {:?}", e)))?;

    let creds = StoredCredentials {
        organization_id: org_id.to_string(),
        session_token: session_token.to_string(),
    };

    let json = serde_json::to_string(&creds)
        .map_err(|e| AppError::Storage(format!("Failed to serialize credentials: {:?}", e)))?;

    entry
        .set_password(&json)
        .map_err(|e| AppError::Storage(format!("Failed to store credentials: {:?}", e)))?;

    Ok(())
}

/// Delete credentials from OS keychain.
pub fn delete_credentials() -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, CREDENTIALS_KEY)
        .map_err(|e| AppError::Storage(format!("Failed to create keyring entry: {:?}", e)))?;

    // Ignore NoEntry errors - credential might not exist
    let _ = entry.delete_credential();

    Ok(())
}

// ============================================================================
// Ollama Credentials
// ============================================================================

/// Load Ollama session token from OS keychain.
/// Returns None if credentials don't exist or on any error.
pub fn load_ollama_credentials() -> Option<String> {
    let entry = Entry::new(SERVICE_NAME, OLLAMA_CREDENTIALS_KEY).ok()?;
    entry.get_password().ok()
}

/// Save Ollama session token to OS keychain.
pub fn save_ollama_credentials(session_token: &str) -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, OLLAMA_CREDENTIALS_KEY)
        .map_err(|e| AppError::Storage(format!("Failed to create keyring entry: {:?}", e)))?;

    entry
        .set_password(session_token)
        .map_err(|e| AppError::Storage(format!("Failed to store Ollama credentials: {:?}", e)))?;

    Ok(())
}

/// Delete Ollama session token from OS keychain.
pub fn delete_ollama_credentials() -> Result<(), AppError> {
    let entry = Entry::new(SERVICE_NAME, OLLAMA_CREDENTIALS_KEY)
        .map_err(|e| AppError::Storage(format!("Failed to create keyring entry: {:?}", e)))?;

    // Ignore NoEntry errors - credential might not exist
    let _ = entry.delete_credential();

    Ok(())
}

File: src/bun/credentials.ts
import { Entry } from "@napi-rs/keyring";
import { AppError } from "./errors";

const SERVICE_NAME = "dev.xikxp1.claude-monitor";
const CREDENTIALS_KEY = "credentials";
const OLLAMA_CREDENTIALS_KEY = "ollama_credentials";

type StoredCredentials = {
  organization_id: string;
  session_token: string;
};

export function loadCredentials(): [string, string] | null {
  try {
    const raw = getPassword(CREDENTIALS_KEY);
    if (!raw) {
      return null;
    }
    const creds = JSON.parse(raw) as StoredCredentials;
    return [creds.organization_id, creds.session_token];
  } catch {
    return null;
  }
}

export function saveCredentials(orgId: string, sessionToken: string): void {
  try {
    setPassword(CREDENTIALS_KEY, JSON.stringify({ organization_id: orgId, session_token: sessionToken }));
  } catch (error) {
    throw new AppError(`Storage error: Failed to store credentials: ${String(error)}`);
  }
}

export function deleteCredentials(): void {
  deletePassword(CREDENTIALS_KEY);
}

export function loadOllamaCredentials(): string | null {
  try {
    return getPassword(OLLAMA_CREDENTIALS_KEY);
  } catch {
    return null;
  }
}

export function saveOllamaCredentials(sessionToken: string): void {
  try {
    setPassword(OLLAMA_CREDENTIALS_KEY, sessionToken);
  } catch (error) {
    throw new AppError(`Storage error: Failed to store Ollama credentials: ${String(error)}`);
  }
}

export function deleteOllamaCredentials(): void {
  deletePassword(OLLAMA_CREDENTIALS_KEY);
}

function getPassword(account: string): string | null {
  return new Entry(SERVICE_NAME, account).getPassword();
}

function setPassword(account: string, password: string): void {
  new Entry(SERVICE_NAME, account).setPassword(password);
}

function deletePassword(account: string): void {
  try {
    new Entry(SERVICE_NAME, account).deletePassword();
  } catch {
    // Clearing credentials should be idempotent from the app's perspective.
  }
}
